internal/storage/storage.go

```diff
@@ -0,0 +1,56 @@
+package storage
+
+import (
+	"time"
+
+	t "github.com/offen/docker-volume-backup/internal/types"
+)
+
+// Interface for defining functions which all storage providers support.
+type Backend interface {
+	Copy(file string) error
+	Prune(deadline time.Time, pruningPrefix string) error
+}
```

Comment:
The fact that we still need package `types` is somehow triggering an itch of mine as it shows we don't have a "one-way flow" of responsibilities in the code. Here's a suggestion about how to maybe get rid of that:

Instead of passing in the `stats` object into the constructor functions, how about changing the interface to look like this:

```go
type Backend interface {
	Copy(file string) (CopyStats, error)
	Prune(deadline time.Time, pruningPrefix string) (PruneStats, error)
}
```

Then, the top level script could merge the returned stats on top of its own values without having to pass around a mutable object. Also we would have a clear graph of package dependencies.

What do you think?

Reply:
I have actually implemented exactly that already, but removed it again.
The problem I saw was, that StorageStats contains fields of Stats for each backend (S3, WebDav, ...).
Inside the `copyArchive` method of `script.go`, you would have to introduce something a switch case to check what backend is being told to copy right now so the stats get stored in the correct field. This felt like an ugly solution to me, as the goal was to remove backend-dependent code from `script.go` and now we would reintroduce some again.
Maybe reorganizing the stats for mitigating this issue would be possible, I didn't use or reverse engineer the stats and their inner workings or how they work in this project.

Reply:
We could do something like this: instead of defining `StoragesStats` as a struct type, we can also use a `map[string]StorageStats`. Then, a `Name() string` method is added to `storage.Backend`. Like this, we wouldn't need a switch case or any backend dependent code and still don't need to pass down script level objects. What do you think about solving it like this?